src/Build/Evaluation/Expander.cs

```diff
@@ -3618,6 +3618,62 @@ private bool TryExecuteWellKnownFunction(out object returnVal, object objectInst
                                 return true;
                             }
                         }
+                        else if (string.Equals(_methodMethodName, nameof(IntrinsicFunctions.GetCurrentToolsDirectory), StringComparison.OrdinalIgnoreCase))
```

Comment:
The number of if checks here is starting to get large. Also I am not sure whether OrdinalIgnoreCase means that strings of different lengths must be compared (?)

If this starts showing up in the profile it you could potentially store all the candidates upper cased, and upper case the test string then make all the tests using Ordinal comparison.

Reply:
Agreed that we should keep an eye out for this in the profiler, and if it does show up we have a few options for perf: length restrictions, prefix matching, generating this code from metadata to build a more-optimal matcher (that sounds pretty fun, actually). But I'm optimistic that this will only help, since reflection is so expensive.

Re: length checks, since the APIs we can fast-path are all ANSI strings (because they're mostly public API from corefx/MSBuild itself, which is all English) we should be able to short-circuit to failure on strings of different length. I'm fine with someone having a slow experience if they try to call `$([System.String]::😬😱🤦🏻‍♂️())`.

Reply:
OrdinalIgnoreCase already bails on unequal lengths: https://github.com/dotnet/coreclr/blob/9c4d496071c2ffdc5a8793835694d1a8ac267522/src/System.Private.CoreLib/shared/System/String.Comparison.cs#L666-L670

IIRC, it is effectively Ordinal with each `char` going through `ToUpperVariant()` before being compared, and ToUpperInvariant() cannot turn one `char` into two.

Reply:
If this is actually as bottleneck, some sort of lookup table on (receiver type, method name), maybe +args, would probably be better. It might also clean up the code.

Reply:
I'm thinking something like 
``` C#
 void RegisterIntrinsicFastPath(Type receiver, string methodName, Func<object> handler) { /* ... */ }
 void RegisterIntrinsicFastPath<TArg0>(Type receiver, string methodName, Func<TArg0, object> handler) { /* ... */}
 void RegisterIntrinsicFastPath<TArg0, TArg1>(Type receiver, string methodName, Func<TArg0, TArg1, object> handler) { /* ... */ }
```

Used like this:
 ```
RegisterIntrinsicFastPath(typeof(string), nameof(string.Equals), (string x, string y) => string.Equals(x, y));
```

This would also be fun to implement, but yeah, unless there's evidence these comparisons are causing a problem, we can leave that for another day.

Reply:
Another option if needed is a hand-tuned switch by length like here
https://github.com/Microsoft/msbuild/issues/3466

I'll mark this resolved though.